### PR TITLE
Spark: Add more tests for MERGE

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR brings more tests for MERGE operations in Spark. In particular, tests with self subqueries, isolation level tests, null handling, updating a single row group in a file with multiple row groups.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @rdblue @dilipbiswal @RussellSpitzer @mehtaashish23

**Comment by aokolnychyi** [comment]
This PR helped to detect the issue fixed in #2215. I'll rebase once merged.

**Comment by aokolnychyi** [comment]
The underlying issue was merged, I rebased this PR as well. It is ready for review.

**Review by dilipbiswal — APPROVED** [review]

**Comment by dilipbiswal** [comment]
@aokolnychyi Nice !!

File: spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java [hunk 1]
```diff
@@ -50,6 +74,361 @@ public void removeTables() {
...
+    createAndInitTable("id INT, v STRING",
+        "{ \"id\": null, \"v\": \"v1\" }\n" +
+        "{ \"id\": 2, \"v\": \"v2\" }");
+
+    createOrReplaceView("source",
+        "{ \"id\": null, \"v\": \"v1_1\" }\n" +
+        "{ \"id\": 4, \"v\": \"v4\" }");
+
+    sql("MERGE INTO %s t USING source " +
+        "ON t.id == source.id " +
```
Comment (rdblue) [inline spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java:264]: I didn't know you could use `==`.

**Review by rdblue — APPROVED** [review]

**Pull request merged**